### ISSUE TITLE
undefined symbol: netsnmp_transport_config_compare -  error

### ISSUE THREAD

**Issue by bijenkins** [fgimian/easysnmp#38]
Python 2.7.10
RHEL 6.7
GCC

Looks to a issue with the way the interface.c file is compiled to interace.so... Any ideas here?

from CLI:

```
Python 2.7.10 (default, Jul 21 2015, 10:07:01) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-11)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> 
>>> from easysnmp import Session
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/JKHY/bijenkins/.pyenv/versions/because-the-internet/lib/python2.7/site-packages/easysnmp/__init__.py", line 1, in <module>
    from .easy import (  # noqa
  File "/home/JKHY/bijenkins/.pyenv/versions/because-the-internet/lib/python2.7/site-packages/easysnmp/easy.py", line 3, in <module>
    from .session import Session
  File "/home/JKHY/bijenkins/.pyenv/versions/because-the-internet/lib/python2.7/site-packages/easysnmp/session.py", line 8, in <module>
    from . import interface
ImportError: /home/JKHY/bijenkins/.pyenv/versions/because-the-internet/lib/python2.7/site-packages/easysnmp/interface.so: undefined symbol: netsnmp_transport_config_compare
>>> 
```

From python:

```
(because-the-internet)[bijenkins@bmograbadev ~]$ pip install easysnmp
Collecting easysnmp
Installing collected packages: easysnmp
Successfully installed easysnmp-0.2.4
(because-the-internet)[bijenkins@bmograbadev ~]$ 
(because-the-internet)[bijenkins@bmograbadev ~]$ 
(because-the-internet)[bijenkins@bmograbadev ~]$ 
(because-the-internet)[bijenkins@bmograbadev ~]$ python easysnmptest.py
Traceback (most recent call last):
  File "easysnmptest.py", line 1, in <module>
    from easysnmp import Session
  File "/home/JKHY/bijenkins/.pyenv/versions/because-the-internet/lib/python2.7/site-packages/easysnmp/__init__.py", line 1, in <module>
    from .easy import (  # noqa
  File "/home/JKHY/bijenkins/.pyenv/versions/because-the-internet/lib/python2.7/site-packages/easysnmp/easy.py", line 3, in <module>
    from .session import Session
  File "/home/JKHY/bijenkins/.pyenv/versions/because-the-internet/lib/python2.7/site-packages/easysnmp/session.py", line 8, in <module>
    from . import interface
ImportError: /home/JKHY/bijenkins/.pyenv/versions/because-the-internet/lib/python2.7/site-packages/easysnmp/interface.so: undefined symbol: netsnmp_transport_config_compare
```

**Comment by nnathan** [comment]
I've not come across RHEL but can you check if interface.so links to the net-snmp library, and which version:

```
$ ldd ~/.virtualenv/default/lib/python2.7/site-packages/easysnmp/interface.so
        linux-vdso.so.1 =>  (0x00007ffeab9e7000)
        libnetsnmp.so.30 => /usr/lib/x86_64-linux-gnu/libnetsnmp.so.30 (0x00007f82d7c79000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f82d7a5b000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f82d7695000)
        libcrypto.so.1.0.0 => /lib/x86_64-linux-gnu/libcrypto.so.1.0.0 (0x00007f82d72ba000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f82d8174000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f82d70b6000)
```

Also, can you tell me which version of net-snmp you're using.

Thanks.

**Comment by fgimian** [comment]
I'll be closing this issue now due to lack of response.  I would suggest that this is a result of the incorrect Net-SNMP version.  Please read the instructions at http://easysnmp.readthedocs.org/en/latest/#installation

**Comment by kamakazikamikaze** [comment]
I hate to dig up old threads but I'm also having this issue on OS X. Please bear with me, this post is going to be all over the place. I'm not 100% certain but I think `virtualenv` and Python 2.7.10 may be part of the issue. Note that I'm using OS X 10.11.4. I am *not* having this issue in Ubuntu 14.10. 

Within the `virtualenv` I could install the `easysnmp` package without error but couldn't import it in my script, seeing the same error printed out as @bijenkins. iPython also displayed the error when trying to import the global package, however upgrading from 2.7.10 to 2.7.11 resolved the issue. The global Python 2.7.11 interpreter seems to have no issue loading the package and running it.

Since OS X does not have `ldd` I ran `otool -L` instead (as suggested [here](https://discussions.apple.com/thread/309193?start=0&tstart=0)) first on the globally installed package which gives the output:

    kentcoble$ otool -L /Users/kentcoble/.python-eggs/easysnmp-0.2.4-py2.7-macosx-10.10-intel.egg-tmp/easysnmp/interface.so
    /Users/kentcoble/.python-eggs/easysnmp-0.2.4-py2.7-macosx-10.10-intel.egg-tmp/easysnmp/interface.so:
    	/opt/local/lib/libnetsnmp.30.dylib (compatibility version 31.0.0, current version 31.3.0)
    	/opt/local/lib/libcrypto.1.0.0.dylib (compatibility version 1.0.0, current version 1.0.0)
    	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1213.0.0)

I ran the command again but this time on the Virtualenv instance:

    (snmpbench) kentcoble$ otool -L ~/Envs/snmpbench/lib/python2.7/site-packages/easysnmp/interface.so 
    /Users/kentcoble/Envs/snmpbench/lib/python2.7/site-packages/easysnmp/interface.so:
	    /usr/lib/libcrypto.0.9.8.dylib (compatibility version 0.9.8, current version 0.9.8)
	    /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)
	    /usr/local/lib/gcc/4.9/libgcc_s.1.dylib (compatibility version 1.0.0, current version 1.0.0)

The `virtualenv` seems to link to the older libraries which may be the root of the issue.

(I noticed that while Net-SNMP 5.7.* is required, it looks like support for 5.6.2.1 was added for OS X. I went ahead anyways and installed 5.7.3 via `brew` and linked it to the system. Reinstalling `easysnmp` in the `virtualenv` did not change anything.)

**Comment by nnathan** [comment]
You may need to verify the extension that comes out as `interface.so` in the virtualenv is linking to the brew installed net-snmp. You can use `otool -L` to verify this.

There are ways to _force_ linkage to net-snmp.

**Comment by kamakazikamikaze** [comment]
I ran that on the `virtualenv` which should be the second `otool -L` output. I don't know why it doesn't have the `libnetsnmp` listed/linked, but I was looking into how to force it. I'll see what I can do over the course of this next week before my finals start.

Has anyone else come across this issue (that you know of) with `virtualenv` and `easysnmp`?

**Comment by nnathan** [comment]
Try doing the `pip install easysnmp` with the following:

```bash
bash-3.2$ export PATH=/usr/local/Cellar/net-snmp/5.7.3/bin:$PATH
bash-3.2$ echo $PATH
/usr/local/Cellar/net-snmp/5.7.3/bin:/usr/local/Cellar/net-snmp/5.7.3/bin:/Users/naveen/e/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
bash-3.2$ which net-snmp-config
/usr/local/Cellar/net-snmp/5.7.3/bin/net-snmp-config
bash-3.2$ net-snmp-config --libs
-L/usr/local/Cellar/net-snmp/5.7.3/lib -lnetsnmp -lcrypto
```

I think this is a good start. However I think `interface.so` will link to the incorrect library. That's just a guess.

**Comment by kamakazikamikaze** [comment]
Ok, found a solution thanks to the help of [this SuperUser thread](http://superuser.com/a/721564/457817). Unfortunately this requires manually linking to the correct libraries, however we may be able to change `setup.py` to do so. It's probably not kosher with what it would take, *but* Net-SNMP isn't the easiest thing to work with on Mac.

In order to correct the libraries associated to `interface.so`, the `install_name_tool` is needed to redirect where it is pointing to.

```
$ install_name_tool -change /usr/lib/libnetsnmp.25.dylib /usr/local/opt/net-snmp/lib/libnetsnmp.30.dylib ../path/to/interface.so

$ install_name_tool -change /usr/lib/libcrypto.0.9.8.dylib /usr/local/opt/openssl/lib/libcrypto.1.0.0.dylib ../path/to/interface.so
```

I can look into finding a way to detect which Net-SNMP is installed if OS X is found to be the target platform and figure out how to set flags to force linking to the 5.7.x version.

**Comment by kamakazikamikaze** [comment]
@nnathan I think I've resolved this in #43. By checking if `brew` is available and the Net-SNMP package has been installed, we can have `brew info <pkg>` spit out the compiler flags and use those instead.

    $ brew info net-snmp
    net-snmp: stable 5.7.3 (bottled) [keg-only]
    Implements SNMP v1, v2c, and v3, using IPv4 and IPv6
    http://www.net-snmp.org/
    /usr/local/Cellar/net-snmp/5.7.3 (538 files, 12.8M) *
      Built from source
    From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/net-snmp.rb
    ==> Dependencies
    Required: openssl ✔
    ==> Options
    --with-python
        Build with python support
    ==> Caveats
    This formula is keg-only, which means it was not symlinked into /usr/local.

    OS X already provides this software and installing another version in
    parallel can cause all kinds of trouble.

    Generally there are no consequences of this for you. If you build your
    own software and it requires this formula, you'll need to add to your
    build variables:

        LDFLAGS:  -L/usr/local/opt/net-snmp/lib
        CPPFLAGS: -I/usr/local/opt/net-snmp/include


This also has to be used on openssl since it's a dependency. Using this method, I've checked that the correct library has been linked by installing and unzipping the `.egg` which gives:

    $ otool -L dist/easysnmp/interface.so 
    easysnmp/interface.so:
	    /usr/local/opt/net-snmp/lib/libnetsnmp.30.dylib (compatibility version 31.0.0, current version 31.3.0)
	    /usr/local/opt/openssl/lib/libcrypto.1.0.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	    /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)

This shouldn't affect users with only 5.6.x installed, however I haven't tested where 5.7.x is installed but the system still points to 5.6.x.
